boa/src/object/gcobject.rs

```diff
@@ -176,31 +206,106 @@ impl GcObject {
+                                has_parameter_expressions || param.init().is_some();
+
+                            let value = match args.get(i).cloned() {
+                                None | Some(Value::Undefined) => param
+                                    .init()
+                                    .map(|init| init.run(context).ok())
+                                    .flatten()
+                                    .unwrap_or_default(),
+                                Some(value) => value,
+                            };
+
+                            function
+                                .add_arguments_to_environment(param, value, &local_env, context);
+                        }
+
+                        if has_parameter_expressions {
+                            // Create a second environment when default parameter expressions are used
+                            // This prevents variables declared in the function body from being
+                            // used in default parameter initializers.
+                            // https://tc39.es/ecma262/#sec-functiondeclarationinstantiation
+                            let second_env = FunctionEnvironmentRecord::new(
+                                this_function_object,
+                                if construct || !flags.is_lexical_this_mode() {
+                                    Some(this)
+                                } else {
+                                    None
+                                },
+                                Some(local_env),
+                                // Arrow functions do not have a this binding https://tc39.es/ecma262/#sec-function-environment-records
+                                if flags.is_lexical_this_mode() {
+                                    BindingStatus::Lexical
+                                } else {
+                                    BindingStatus::Uninitialized
+                                },
+                                Value::undefined(),
+                            );
+                            context.push_environment(second_env);
+                        }
 
                         FunctionBody::Ordinary(body.clone())
                     }
                 }
-            } else {
-                return context.throw_type_error("function object is not callable");
             }
         } else {
             return context.throw_type_error("not a function");
         };
 
-        match f_body {
-            FunctionBody::BuiltInFunction(func) => func(this, args, context),
-            FunctionBody::BuiltInConstructor(func) => func(&Value::undefined(), args, context),
+        match body {
+            FunctionBody::BuiltInConstructor(function) if construct => {
+                function(&this_target, args, context)
+            }
+            FunctionBody::BuiltInFunction(_) if construct => {
+                unreachable!("Cannot have a function in construct")
+            }
```

Comment:
Are these lines necessary? Since we know this combination (`BuiltInFuncion` and `construct`) will never happen can't we just assume it won't happen and leave only line 278 to deal with `BuiltInFuncion` bodies?

Reply:
You're right, since the condition on `construct` on line 141 would make it always return `BuiltInConstructor`.